Index: busybox-1.27.2/networking/ifupdown.c
===================================================================
--- busybox-1.27.2.orig/networking/ifupdown.c
+++ busybox-1.27.2/networking/ifupdown.c
@@ -1506,6 +1506,9 @@ int ifupdown_main(int argc UNUSED_PARAM,
 				}
 				state = state->link;
 			}
+
+            fsync(fileno(new_state_fp));
+
 			fclose(new_state_fp);
 			xrename(IFSTATE_FILE_PATH".new", IFSTATE_FILE_PATH);
 			llist_free(state_list, free);
