From: Marc Kleine-Budde <mkl@pengutronix.de>
Date: Sun, 10 Jan 2010 01:10:34 +0100
Subject: [PATCH] xslt-config: add SYSROOT support

Signed-off-by: Marc Kleine-Budde <mkl@pengutronix.de>
---
 xslt-config.in |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xslt-config.in b/xslt-config.in
index 45c3e28f2f12..5c009c4f4ad2 100644
--- a/xslt-config.in
+++ b/xslt-config.in
@@ -1,6 +1,6 @@
 #! /bin/sh
 
-prefix=@prefix@
+prefix=${SYSROOT}@prefix@
 exec_prefix=@exec_prefix@
 exec_prefix_set=no
 includedir=@includedir@
