From: Jobst Wellensiek <jobst.wellensiek@wago.com>
Date: Thu, 3 Mar 2022 10:42:48 +0100
Subject: [PATCH] set default stack size to 1Mb

---
 src/core/lib/gprpp/thd.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/core/lib/gprpp/thd.h b/src/core/lib/gprpp/thd.h
index c4ce04cec460..ae189d4cd67a 100644
--- a/src/core/lib/gprpp/thd.h
+++ b/src/core/lib/gprpp/thd.h
@@ -47,7 +47,7 @@ class Thread {
  public:
   class Options {
    public:
-    Options() : joinable_(true), tracked_(true), stack_size_(0) {}
+    Options() : joinable_(true), tracked_(true), stack_size_(1024 * 1000) {}
     /// Set whether the thread is joinable or detached.
     Options& set_joinable(bool joinable) {
       joinable_ = joinable;
