From f2605394fcf5357ce538af954f8bb4b17205f598 Mon Sep 17 00:00:00 2001
From: Timur Celik <timur.celik@wago.com>
Date: Thu, 6 Mar 2014 16:07:34 +0100
Subject: [PATCH] always use priority inheritance capable mutex, since we are
 using rtlinux

---
 glib/gthread-posix.c | 11 +++--------
 1 file changed, 3 insertions(+), 8 deletions(-)

Index: glib-2.56.3/glib/gthread-posix.c
===================================================================
--- glib-2.56.3.orig/glib/gthread-posix.c
+++ glib-2.56.3/glib/gthread-posix.c
@@ -85,29 +85,24 @@
 static pthread_mutex_t *
 g_mutex_impl_new (void)
 {
-  pthread_mutexattr_t *pattr = NULL;
   pthread_mutex_t *mutex;
   gint status;
-#ifdef PTHREAD_ADAPTIVE_MUTEX_INITIALIZER_NP
   pthread_mutexattr_t attr;
-#endif
 
   mutex = malloc (sizeof (pthread_mutex_t));
   if G_UNLIKELY (mutex == NULL)
     g_thread_abort (errno, "malloc");
 
-#ifdef PTHREAD_ADAPTIVE_MUTEX_INITIALIZER_NP
   pthread_mutexattr_init (&attr);
+#ifdef PTHREAD_ADAPTIVE_MUTEX_INITIALIZER_NP
   pthread_mutexattr_settype (&attr, PTHREAD_MUTEX_ADAPTIVE_NP);
-  pattr = &attr;
 #endif
 
-  if G_UNLIKELY ((status = pthread_mutex_init (mutex, pattr)) != 0)
+  pthread_mutexattr_setprotocol(&attr, PTHREAD_PRIO_INHERIT);
+  if G_UNLIKELY ((status = pthread_mutex_init (mutex, &attr)) != 0)
     g_thread_abort (status, "pthread_mutex_init");
 
-#ifdef PTHREAD_ADAPTIVE_MUTEX_INITIALIZER_NP
   pthread_mutexattr_destroy (&attr);
-#endif
 
   return mutex;
 }
