diff -ur strongswan-5.3.5.orig/src/ipsec/_ipsec.in strongswan-5.3.5/src/ipsec/_ipsec.in
--- strongswan-5.3.5.orig/src/ipsec/_ipsec.in	2019-02-11 17:20:27.492999153 +0100
+++ strongswan-5.3.5/src/ipsec/_ipsec.in	2019-02-12 07:28:20.941713172 +0100
@@ -257,7 +258,8 @@
 			loop=110
 			while [ $loop -gt 0 ] ; do
 				kill -0 $spid 2>/dev/null || break
-				sleep 0.1 2>/dev/null
+				#sleep 0.1 2>/dev/null    !!doesn't work with busybox!!
+				usleep 100000
 				if [ $? -ne 0 ]
 				then
 					sleep 1
