From 76bd6d41f277af0e83568afed77dbd53102bf422 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Mon, 14 Nov 2016 15:10:42 +0000
Subject: [PATCH] serial: option: add EC25 and EC21

---
 drivers/usb/serial/option.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/usb/serial/option.c b/drivers/usb/serial/option.c
index b8c83a3..f546eff 100644
--- a/drivers/usb/serial/option.c
+++ b/drivers/usb/serial/option.c
@@ -573,6 +573,8 @@ static void option_instat_callback(struct urb *urb);
 static const struct usb_device_id option_ids[] = {
 	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_UC20) },
 	{ USB_DEVICE(QUECTEL_VENDOR_ID, QUECTEL_PRODUCT_EC20) },
+        { USB_DEVICE(0x2c7c, 0x0125) }, /* Quectel EC25, EC20 R2.0  Mini PCIe */
+        { USB_DEVICE(0x2c7c, 0x0121) }, /* Quectel EC21 Mini PCIe */
 	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_COLT) },
 	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_RICOLA) },
 	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_RICOLA_LIGHT) },
-- 
2.7.4

