From 8d7e2183a7c933c9d11be3eda0b87486724c8a98 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Tue, 24 Jan 2017 11:55:38 +0000
Subject: [PATCH] tps65218: fix wrong values and adjust to LS2

---
 drivers/regulator/tps65218-regulator.c | 23 ++++++++++++++++++-----
 1 file changed, 18 insertions(+), 5 deletions(-)

diff --git a/drivers/regulator/tps65218-regulator.c b/drivers/regulator/tps65218-regulator.c
index 2877640..5fff734 100644
--- a/drivers/regulator/tps65218-regulator.c
+++ b/drivers/regulator/tps65218-regulator.c
@@ -230,12 +230,14 @@ static struct regulator_ops tps65218_ldo1_dcdc34_ops = {
 	.set_suspend_disable	= tps65218_pmic_set_suspend_disable,
 };
 
-static const int ls3_currents[] = { 100, 200, 500, 1000 };
+/* values in uA */
+static const int ls3_currents[] = { 100000, 200000, 500000, 1000000 };
 
 static int tps65218_pmic_set_input_current_lim(struct regulator_dev *dev,
 					       int lim_uA)
 {
 	unsigned int index = 0;
+	int shift = 0;
 	unsigned int num_currents = ARRAY_SIZE(ls3_currents);
 	struct tps65218 *tps = rdev_get_drvdata(dev);
 
@@ -245,18 +247,22 @@ static int tps65218_pmic_set_input_current_lim(struct regulator_dev *dev,
 	if (index == num_currents)
 		return -EINVAL;
 
+	if (dev->desc->id == TPS65218_LS_3)
+		shift = 2;
+
 	return tps65218_set_bits(tps, dev->desc->csel_reg, dev->desc->csel_mask,
-				 index << 2, TPS65218_PROTECT_L1);
+				 index << shift, TPS65218_PROTECT_L1);
 }
 
 static int tps65218_pmic_set_current_limit(struct regulator_dev *dev,
 					   int min_uA, int max_uA)
 {
 	int index = 0;
+	int shift = 0;
 	unsigned int num_currents = ARRAY_SIZE(ls3_currents);
 	struct tps65218 *tps = rdev_get_drvdata(dev);
 
-	while (index < num_currents && ls3_currents[index] < max_uA)
+	while (index < num_currents && ls3_currents[index] <= max_uA)
 		index++;
 
 	index--;
@@ -264,13 +270,17 @@ static int tps65218_pmic_set_current_limit(struct regulator_dev *dev,
 	if (index < 0 || ls3_currents[index] < min_uA)
 		return -EINVAL;
 
+	if (dev->desc->id == TPS65218_LS_3)
+		shift = 2;
+
 	return tps65218_set_bits(tps, dev->desc->csel_reg, dev->desc->csel_mask,
-				 index << 2, TPS65218_PROTECT_L1);
+				 index << shift, TPS65218_PROTECT_L1);
 }
 
 static int tps65218_pmic_get_current_limit(struct regulator_dev *dev)
 {
 	int retval;
+	int shift = 0;
 	unsigned int index;
 	struct tps65218 *tps = rdev_get_drvdata(dev);
 
@@ -278,7 +288,10 @@ static int tps65218_pmic_get_current_limit(struct regulator_dev *dev)
 	if (retval < 0)
 		return retval;
 
-	index = (index & dev->desc->csel_mask) >> 2;
+	if (dev->desc->id == TPS65218_LS_3)
+		shift = 2;
+
+	index = (index & dev->desc->csel_mask) >> shift;
 
 	return ls3_currents[index];
 }
-- 
2.7.4

