From 9625a911a9dba63ce8259a3b05bb74e174f0f0eb Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Tue, 2 Oct 2018 10:28:53 +0000
Subject: [PATCH 247/292] musb: am35x: add a cast to remove build warning

Signed-off-by: Heinrich Toews <heinrich.toews@wago.com>
---
 drivers/usb/musb/am35x.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/usb/musb/am35x.c b/drivers/usb/musb/am35x.c
index 32d4704..530a1da 100644
--- a/drivers/usb/musb/am35x.c
+++ b/drivers/usb/musb/am35x.c
@@ -469,7 +469,7 @@ static const struct platform_device_info am35x_dev_info = {
 static int of_am35x_probe(struct device_node *np,
 			  struct musb_hdrc_platform_data *pdata)
 {
-	struct musb_hdrc_config *config = pdata->config;
+	struct musb_hdrc_config *config = (struct musb_hdrc_config *) pdata->config;
 	struct omap_musb_board_data *data = (struct omap_musb_board_data *)pdata->board_data;
 
 	/* we presume that pdata was already prepared in pdata-quirks.c */
-- 
2.7.4

