From 4c9005c90c09162c1042644b32bf5584e064c3cb Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Mon, 15 Oct 2018 09:10:01 +0200
Subject: [PATCH] ARM: dts: fix indents

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 arch/arm/boot/dts/Makefile | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index f82081c..a5abce7 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -486,7 +486,7 @@ dtb-$(CONFIG_ARCH_OMAP3) += \
 	am3517-evm.dtb \
 	am3517_mt_ventoux.dtb \
 	am3505-pfc-750_8202.dtb \
-        am3505-pfc-750_8203.dtb \
+	am3505-pfc-750_8203.dtb \
 	am3505-pfc-750_8204.dtb \
 	am3505-pfc-750_8206.dtb \
 	am3505-pfc-750_8207.dtb \
@@ -568,8 +568,8 @@ dtb-$(CONFIG_SOC_AM33XX) += \
 	am335x-sbc-t335.dtb \
 	am335x-sl50.dtb \
 	am335x-wega-rdk.dtb \
-        am335x-pfc-750_8102.dtb \
-        am335x-pfc-750_8100_8101.dtb \
+	am335x-pfc-750_8102.dtb \
+	am335x-pfc-750_8100_8101.dtb \
 	am335x-pfc-750_8208.dtb \
 	am335x-pfc-750_8211.dtb \
 	am335x-pfc-750_8212.dtb \
@@ -579,7 +579,7 @@ dtb-$(CONFIG_SOC_AM33XX) += \
 	am335x-pfc-750_8216.dtb \
 	am335x-pfc-750_8217.dtb \
 	am335x-src-2850_3111_3112.dtb \
-        am335x-pfc100-setupfw.dtb \
+	am335x-pfc100-setupfw.dtb \
 	am335x-pfc200v2-setupfw.dtb \
 	am335x-pfc200v3-setupfw.dtb \
 	am335x-src-setupfw.dtb
-- 
2.7.4

