From 9187b5398a3586bae3656f85815fe73fa2826b84 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Tue, 2 Apr 2019 15:22:53 +0200
Subject: [PATCH 292/292] add new version w02.04.00_02+2

Signed-off-by: Heinrich Toews <heinrich.toews@wago.com>
---
 localversion-wago | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-wago b/localversion-wago
index e6a06ac..ece0a4a 100644
--- a/localversion-wago
+++ b/localversion-wago
@@ -1 +1 @@
--w02.03.00_02+2
+-w02.04.00_02+2
-- 
2.7.4

