From 9cbcafea13038b83cbfc58b2bd59d43db0d7f621 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Tue, 14 Jan 2020 10:55:53 +0100
Subject: [PATCH] add new version w02.06.01

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 localversion-wago | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-wago b/localversion-wago
index a789196..fd26e24 100644
--- a/localversion-wago
+++ b/localversion-wago
@@ -1 +1 @@
-w02.06.00
+w02.06.01
-- 
2.7.4

