From a575a3ea29d0bd174c223a87558eeff309b3ecd9 Mon Sep 17 00:00:00 2001
From: Oliver Rohe <Oliver.Rohe@wago.com>
Date: Wed, 15 Jan 2020 09:43:11 +0100
Subject: [PATCH] driver: char: rmd: remove obsolete sysfs calls

---
 drivers/char/rmd.c | 8 --------
 1 file changed, 8 deletions(-)

diff --git a/drivers/char/rmd.c b/drivers/char/rmd.c
index f33702b..2afebd0 100644
--- a/drivers/char/rmd.c
+++ b/drivers/char/rmd.c
@@ -1032,12 +1032,6 @@ static int rmd_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, rmd);
 	dev_set_drvdata(rmd->dev, rmd);
 
-	ret = device_create_file(rmd->dev, &dev_attr_reset);
-	if (ret) {
-		dev_err(parent, "could not create device attribute\n");
-		return ret;
-	}
-
 	return 0;
 }
 
@@ -1045,8 +1039,6 @@ static int rmd_remove(struct platform_device *pdev)
 {
 	struct rmd *rmd = platform_get_drvdata(pdev);
 
-	device_remove_file(rmd->dev, &dev_attr_reset);
-
 	device_destroy(rmd_base.class, MKDEV(rmd_base.major, rmd->minor));
 	rmd_free_minor(rmd);
 
-- 
2.7.4

