From 8b00c10b924e1055e0b747cbca7c75ba4e5d7fc2 Mon Sep 17 00:00:00 2001
From: "Karfich, Oleg" <Oleg.Karfich@wago.com>
Date: Wed, 18 Mar 2020 09:53:22 +0000
Subject: [PATCH] drivers: char: rmd: irq: be robust against defered probe
 requests

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 drivers/char/rmd.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/char/rmd.c b/drivers/char/rmd.c
index 2afebd0..fa578cb 100644
--- a/drivers/char/rmd.c
+++ b/drivers/char/rmd.c
@@ -942,7 +942,12 @@ static int of_rmd(struct device *dev, struct rmd *rmd)
 	if (ret)
 		return ret;
 
-	rmd->irq = irq_of_parse_and_map(np, 0);
+	ret = of_irq_get(np, 0);
+	if (ret < 0)
+		return ret;
+
+	rmd->irq = ret;
+
 	ret = devm_request_irq(dev, rmd->irq, rmd_isr, IRQF_SHARED |
 						IRQF_THREAD_TBL_LOOKUP |
 						IRQ_TYPE_EDGE_RISING,
-- 
2.7.4

