From 6b7287e0071f3a7da4c766f6ffaaaf1001063e0a Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Thu, 12 Mar 2020 15:22:32 +0100
Subject: [PATCH] add new version w02.08.00

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 localversion-wago | 1 +
 1 file changed, 1 insertion(+)
 create mode 100644 localversion-wago

diff --git a/localversion-wago b/localversion-wago
new file mode 100644
index 00000000..91c1319
--- /dev/null
+++ b/localversion-wago
@@ -0,0 +1 @@
+w02.08.00
-- 
2.7.4

