From 01ee5ff93efa6f754849422b37a14a9fa13fff79 Mon Sep 17 00:00:00 2001
From: Ralf Gliese <gliese@elrest.de>
Date: Tue, 2 Jul 2019 11:46:16 +0200
Subject: [PATCH] serial: imx: rs485: don't enable RTS signal if tx-buffer is
 empty

Signed-off-by: Ralf Gliese <gliese@elrest.de>
---
 drivers/tty/serial/imx.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/tty/serial/imx.c b/drivers/tty/serial/imx.c
index 0340c94..bf534b8 100644
--- a/drivers/tty/serial/imx.c
+++ b/drivers/tty/serial/imx.c
@@ -584,10 +584,13 @@ static void imx_start_tx(struct uart_port *port)
 
 	if (port->rs485.flags & SER_RS485_ENABLED) {
 		temp = readl(port->membase + UCR2);
-		if (port->rs485.flags & SER_RS485_RTS_ON_SEND)
-			imx_port_rts_active(sport, &temp);
-		else
-			imx_port_rts_inactive(sport, &temp);
+		if (!uart_circ_empty(&port->state->xmit) &&
+		    !uart_tx_stopped(port)) {
+			if (port->rs485.flags & SER_RS485_RTS_ON_SEND)
+				imx_port_rts_active(sport, &temp);
+			else
+				imx_port_rts_inactive(sport, &temp);
+		}
 		if (!(port->rs485.flags & SER_RS485_RX_DURING_TX))
 			temp &= ~UCR2_RXEN;
 		writel(temp, port->membase + UCR2);
-- 
2.7.4

