From 7f79a57e01902d81d4c52bf4795496d9e636dcec Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Thu, 8 Oct 2020 08:22:53 +0200
Subject: [PATCH] add new version w02.11.1 in branch maint-fw18 for
 SVN-Release-branch

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 localversion-wago | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-wago b/localversion-wago
index 0cbe931..7d7ca0b 100644
--- a/localversion-wago
+++ b/localversion-wago
@@ -1 +1 @@
-w02.11.00
+w02.11.01
-- 
2.7.4

