From 5813f649a0691d76adc9912f010e2f9d4ef83c47 Mon Sep 17 00:00:00 2001
From: Christian Hohnstaedt <Christian.Hohnstaedt@wago.com>
Date: Thu, 19 Nov 2020 08:22:04 +0100
Subject: [PATCH] add new version w02.11.2 in branch maint-fw18 for
 SVN-Release-branch

Signed-off-by: Christian Hohnstaedt <Christian.Hohnstaedt@wago.com>
---
 localversion-wago | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-wago b/localversion-wago
index 7d7ca0b..0b1dde1 100644
--- a/localversion-wago
+++ b/localversion-wago
@@ -1 +1 @@
-w02.11.01
+w02.11.02
-- 
2.7.4

