Index: net-snmp-5.7.3/snmplib/mib.c
===================================================================

Author: Hans-Florian Scholz, 26.02.2015

--- net-snmp-5.7.3.orig/snmplib/mib.c
+++ net-snmp-5.7.3/snmplib/mib.c
@@ -1440,6 +1440,13 @@ sprint_realloc_uinteger(u_char ** buf, s
         }
     }
 
+    if (!netsnmp_ds_get_boolean(NETSNMP_DS_LIBRARY_ID, NETSNMP_DS_LIB_QUICK_PRINT)) {
+        if (!snmp_strcat(buf, buf_len, out_len, allow_realloc,
+                         (const u_char *) "UINTEGER: ")) {
+            return 0;
+        }
+    }
+
     if (enum_string == NULL ||
         netsnmp_ds_get_boolean(NETSNMP_DS_LIBRARY_ID, NETSNMP_DS_LIB_PRINT_NUMERIC_ENUM)) {
         if (hint) {
