--- a/include/nlohmann/detail/input/json_sax.hpp
+++ b/include/nlohmann/detail/input/json_sax.hpp
@@ -281,6 +281,10 @@ class json_sax_dom_parser
         return errored;
     }
 
+  protected:
+    /// stack to model hierarchy of values
+    std::vector<BasicJsonType*> ref_stack {};
+
   private:
     /*!
     @invariant If the ref stack is empty, then the passed value will be the new
@@ -314,8 +318,6 @@ class json_sax_dom_parser
 
     /// the parsed JSON value
     BasicJsonType& root;
-    /// stack to model hierarchy of values
-    std::vector<BasicJsonType*> ref_stack {};
     /// helper to hold the reference for the next object element
     BasicJsonType* object_element = nullptr;
     /// whether a syntax error occurred
