From cff126b2e24bfa416a056997692a1165b578916e Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Wed, 8 Aug 2018 11:23:11 +0200
Subject: [PATCH] drivers: misc: bootmatrix: set leds in specific states

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 drivers/misc/bootmatrix.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/misc/bootmatrix.c b/drivers/misc/bootmatrix.c
index 072e364..e7f7156 100644
--- a/drivers/misc/bootmatrix.c
+++ b/drivers/misc/bootmatrix.c
@@ -26,6 +26,7 @@
 #include <linux/types.h>
 #include <linux/bitmap.h>
 #include <bootmatrix.h>
+#include <led_mode.h>
 
 #define MAX_KEY_BITS 8
 
@@ -106,6 +107,8 @@ static enum bootmatrix_mode bootmatrix_detect_mode(struct bootmatrix
 	bootmatrix_get_current_key_values(bootmatrix);
 	if (test_bit(RST, matrix) && (test_bit(RESET, matrix) || single_key)) {
 
+		led_mode_static("green", true);
+
 		start = get_time_ns();
 		while (!exit) {
 
@@ -131,6 +134,7 @@ static enum bootmatrix_mode bootmatrix_detect_mode(struct bootmatrix
 								      SECOND)) {
 					mode = PRODUCTION;
 					state = SINGLE_KEY_MODE_STAGE2;
+					led_mode_static("green", false);
 				}
 				break;
 
-- 
2.7.4

