From 7ddb6c0fa1b6472c46a722d6d3a64e39c40b6262 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Tue, 21 Aug 2018 14:38:56 +0200
Subject: [PATCH] drivers: misc: xsection: add blidx field

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 drivers/misc/xsection.c | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/drivers/misc/xsection.c b/drivers/misc/xsection.c
index e0894e0..27b87fb 100644
--- a/drivers/misc/xsection.c
+++ b/drivers/misc/xsection.c
@@ -35,6 +35,7 @@ struct xsection {
 	struct xsection_item devconf;
 	struct xsection_item ethaddr1;
 	struct xsection_item ethaddr2;
+	struct xsection_item blidx;
 };
 
 static struct xsection *xsection;
@@ -216,6 +217,37 @@ static int xsection_param_set_devconf(struct param_d *p, void *priv)
 	return ret < 0 ? ret : 0;
 }
 
+static int xsection_param_get_blidx(struct param_d *p, void *priv)
+{
+	struct xsection_item *xitem = priv;
+	int ret;
+	u8 buf;
+
+	ret = xsection_access_item(xitem, &buf, O_RDONLY);
+	if (ret < 0)
+		return ret;
+
+	free(xitem->value_param);
+	xitem->value_param = basprintf("%d", buf);
+
+	return 0;
+}
+
+static int xsection_param_set_blidx(struct param_d *p, void *priv)
+{
+	struct xsection_item *xitem = priv;
+	int ret;
+	u8 value;
+
+	if (!strcmp(xitem->value_param, ""))
+		return -EINVAL;
+
+	value = simple_strtoul(xitem->value_param, NULL, 0);
+	ret = xsection_access_item(xitem, &value, O_WRONLY);
+
+	return ret < 0 ? ret : 0;
+}
+
 static int xsection_add_dev_param(struct device_node *np,
 				struct xsection *xsection,
 				struct xsection_item *xitem, char *param_name,
@@ -304,6 +336,13 @@ static int xsection_probe(struct device_d *dev)
 	if (ret)
 		goto out_unreg;
 
+	ret = xsection_add_dev_param(np, xsection, &xsection->blidx,
+					"blidx",
+					 xsection_param_set_blidx,
+					 xsection_param_get_blidx);
+	if (ret)
+		goto out_unreg;
+
 	return 0;
 
 out_unreg:
-- 
2.7.4

