From e7ea7dcf3ba9007a90c2d3f240e0a893019bfc4b Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Mon, 27 Aug 2018 16:09:19 +0200
Subject: [PATCH] i2c-omap: read scheme reg only if it is available

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 drivers/i2c/busses/i2c-omap.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/busses/i2c-omap.c b/drivers/i2c/busses/i2c-omap.c
index bdb34ca..96e90e2 100644
--- a/drivers/i2c/busses/i2c-omap.c
+++ b/drivers/i2c/busses/i2c-omap.c
@@ -1121,7 +1121,10 @@ i2c_omap_probe(struct device_d *pdev)
 	 * Also since the omap_i2c_read_reg uses reg_map_ip_* a
 	 * raw_readw is done.
 	 */
-	rev = __raw_readw(i2c_omap->base + 0x04);
+	if (!of_machine_is_compatible("ti,omap3") && !of_machine_is_compatible("ti,omap2"))
+		rev = __raw_readw(i2c_omap->base + 0x04);
+	else
+		rev = 0;
 
 	i2c_omap->scheme = OMAP_I2C_SCHEME(rev);
 	switch (i2c_omap->scheme) {
-- 
2.7.4

