From d7485fa319060038bb6e2b210cb2bb48c2d1850b Mon Sep 17 00:00:00 2001
From: Andreas Schmidt <andreas.schmidt@wago.com>
Date: Thu, 11 Oct 2018 15:08:24 +0200
Subject: [PATCH] gitignore: add m-arm to ignored list

Signed-off-by: Andreas Schmidt <andreas.schmidt@wago.com>
---
 .gitignore | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/.gitignore b/.gitignore
index d83d317..bee9759 100644
--- a/.gitignore
+++ b/.gitignore
@@ -79,3 +79,5 @@ GTAGS
 
 # Kconfig presets
 all.config
+
+m-arm
-- 
2.7.4

