From 06b9263054bca0fe5691132be93f28864513bd25 Mon Sep 17 00:00:00 2001
From: Andreas Schmidt <andreas.schmidt@wago.com>
Date: Thu, 11 Oct 2018 15:10:44 +0200
Subject: [PATCH 111/144] commands: miitool: remove unsupported options a and b

Signed-off-by: Andreas Schmidt <andreas.schmidt@wago.com>
---
 commands/miitool.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/commands/miitool.c b/commands/miitool.c
index dea4f85..8872469 100644
--- a/commands/miitool.c
+++ b/commands/miitool.c
@@ -281,12 +281,6 @@ static int do_miitool(int argc, char *argv[])
 
 	while ((opt = getopt(argc, argv, "vs:r:")) > 0) {
 		switch (opt) {
-		case 'a':
-			addr = simple_strtol(optarg, NULL, 0);
-			break;
-		case 'b':
-			bus = simple_strtoul(optarg, NULL, 0);
-			break;
 		case 's':
 			action = MIITOOL_SHOW;
 			phydevname = xstrdup(optarg);
-- 
2.7.4

