From a806be4275de6bd368c911f28632ac2846279506 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Wed, 17 Oct 2018 07:37:31 +0200
Subject: [PATCH] gitignore: ignore update-svn-patches.sh

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 .gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/.gitignore b/.gitignore
index bee9759..dd57494 100644
--- a/.gitignore
+++ b/.gitignore
@@ -81,3 +81,4 @@ GTAGS
 all.config
 
 m-arm
+update-svn-patches.sh
-- 
2.7.4

