From 49acca3be5ac166c69cbb3c85cd7606981ca94f8 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Tue, 16 Oct 2018 10:32:43 +0000
Subject: [PATCH] net: davinci-emac: add device tree support

Signed-off-by: Heinrich Toews <heinrich.toews@wago.com>
---
 drivers/net/davinci_emac.c | 30 ++++++++++++++++++++++++++++--
 1 file changed, 28 insertions(+), 2 deletions(-)

diff --git a/drivers/net/davinci_emac.c b/drivers/net/davinci_emac.c
index 75a917a..10014b7 100644
--- a/drivers/net/davinci_emac.c
+++ b/drivers/net/davinci_emac.c
@@ -514,6 +514,19 @@ out:
 	return ret;
 }
 
+struct davinci_emac_platform_data *davinci_emac_probe_dt(struct device_node *np)
+{
+	struct davinci_emac_platform_data *pdata;
+
+	pdata = xzalloc(sizeof(*pdata));
+
+	of_property_read_u32(np, "phy-addr", &pdata->phy_addr);
+	pdata->force_link = of_property_read_bool(np, "force-link");
+	pdata->interface_rmii = of_property_read_bool(np, "interface-rmii");
+
+	return pdata;
+}
+
 static int davinci_emac_probe(struct device_d *dev)
 {
 	struct resource *iores;
@@ -524,11 +537,15 @@ static int davinci_emac_probe(struct device_d *dev)
 
 	dev_dbg(dev, "+ emac_probe\n");
 
-	if (!dev->platform_data) {
+	if (dev->device_node)
+		pdata = davinci_emac_probe_dt(dev->device_node);
+	else
+		pdata = dev->platform_data;
+
+	if (!pdata) {
 		dev_err(dev, "no platform_data\n");
 		return -ENODEV;
 	}
-	pdata = dev->platform_data;
 
 	priv = xzalloc(sizeof(*priv));
 	dev->priv = priv;
@@ -620,9 +637,18 @@ static void davinci_emac_remove(struct device_d *dev)
 	davinci_emac_halt(&priv->edev);
 }
 
+static __maybe_unused struct of_device_id davinci_emac_dt_ids[] = {
+	{
+		.compatible = "ti,am3517-emac",
+	}, {
+		/* sentinel */
+	}
+};
+
 static struct driver_d davinci_emac_driver = {
 	.name   = "davinci_emac",
 	.probe  = davinci_emac_probe,
 	.remove = davinci_emac_remove,
+	.of_compatible = DRV_OF_COMPAT(davinci_emac_dt_ids),
 };
 device_platform_driver(davinci_emac_driver);
-- 
2.7.4

