From ab75c4e6c38e3d3b88f53a06709f02a296e85099 Mon Sep 17 00:00:00 2001
From: Boerje Sewing <Boerje.Sewing@who-ing.de>
Date: Fri, 25 Jan 2019 14:00:31 +0100
Subject: [PATCH] Version bump to w03.00.01_05

Signed-off-by: Boerje Sewing <Boerje.Sewing@who-ing.de>
---
 localversion-wago | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-wago b/localversion-wago
index 1dfb5c5c3bb..b30311e5058 100644
--- a/localversion-wago
+++ b/localversion-wago
@@ -1 +1 @@
--w03.00.00_05
+-w03.00.01_05
-- 
2.20.1

