From 2d12c9079b60c91be48176900f9bc323fb848ad2 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Thu, 28 Nov 2019 10:27:03 +0100
Subject: [PATCH] arm: omap: pfc-xload: select PARAMETER config

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 arch/arm/mach-omap/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/arm/mach-omap/Kconfig b/arch/arm/mach-omap/Kconfig
index 9c93007..29267a0 100644
--- a/arch/arm/mach-omap/Kconfig
+++ b/arch/arm/mach-omap/Kconfig
@@ -226,6 +226,7 @@ if MACH_WAGO_PFC_AM335X || MACH_WAGO_PFC_AM35XX
 
 config WAGO_PFC_XLOAD
 	bool
+	select PARAMETER
 	prompt "Activate Wago PFC specific xload startup"
 	help
 	  Say Y to enable pfc specific xload startup code. This simultaneously
-- 
2.7.4

