From d053a21db88fc97f80c14eb021f0dc841b158582 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Fri, 22 Nov 2019 12:35:54 +0100
Subject: [PATCH] arm: dts: pfc: add emmc evironment node

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 arch/arm/dts/am335x-pfc.dtsi | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/arch/arm/dts/am335x-pfc.dtsi b/arch/arm/dts/am335x-pfc.dtsi
index c88361a..7515c10 100644
--- a/arch/arm/dts/am335x-pfc.dtsi
+++ b/arch/arm/dts/am335x-pfc.dtsi
@@ -40,6 +40,13 @@
 			file-path = "barebox.env";
 			status = "disabled";
 		};
+
+		environment_emmc: environment-emmc {
+			compatible = "barebox,environment";
+			device-path = &mmc2, "partname:0";
+			file-path = "barebox.env";
+			status = "disabled";
+		};
 	};
 
 	memory {
-- 
2.7.4

