From 9286537a4cb918ae3b32c94022fbc23600792872 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Tue, 11 Feb 2020 14:06:59 +0100
Subject: [PATCH] drivers: misc: xsection: add boot_mode_id_ext field

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 drivers/misc/xsection.c | 19 +++++++++++++++++++
 include/xsection.h      |  3 +++
 2 files changed, 22 insertions(+)

diff --git a/drivers/misc/xsection.c b/drivers/misc/xsection.c
index 26db6e0..cafda6f 100644
--- a/drivers/misc/xsection.c
+++ b/drivers/misc/xsection.c
@@ -32,6 +32,7 @@ struct xsection_item {
 struct xsection {
 	struct device_d dev;
 	struct xsection_item boot_mode_id;
+	struct xsection_item boot_mode_id_ext;
 	struct xsection_item devconf;
 	struct xsection_item ethaddr1;
 	struct xsection_item ethaddr2;
@@ -72,6 +73,17 @@ int xsection_set_boot_mode_id(u8 *value)
 							O_WRONLY) : -ENODEV;
 }
 
+int xsection_get_boot_mode_id_ext(u8 *buf)
+{
+	return xsection ? xsection_access_item(&xsection->boot_mode_id_ext, buf,
+							O_RDONLY) : -ENODEV;
+}
+
+int xsection_set_boot_mode_id_ext(u8 *value)
+{
+	return xsection ? xsection_access_item(&xsection->boot_mode_id_ext, value,
+							O_WRONLY) : -ENODEV;
+}
 int xsection_get_devconf(u16 *buf)
 {
 	return xsection ? xsection_access_item(&xsection->devconf, buf,
@@ -322,6 +334,13 @@ static int xsection_probe(struct device_d *dev)
 	if (ret)
 		goto out_unreg;
 
+	ret = xsection_add_dev_param(np, xsection, &xsection->boot_mode_id_ext,
+					"boot_mode_id_ext",
+					xsection_param_set_boot_mode_id,
+					xsection_param_get_boot_mode_id);
+	if (ret)
+		goto out_unreg;
+
 	ret = xsection_add_dev_param(np, xsection, &xsection->devconf,
 					"devconf",
 					xsection_param_set_devconf,
diff --git a/include/xsection.h b/include/xsection.h
index 3947b24..dfb3708 100644
--- a/include/xsection.h
+++ b/include/xsection.h
@@ -4,6 +4,9 @@
 int xsection_get_boot_mode_id(u8 *buf);
 int xsection_set_boot_mode_id(u8 *value);
 
+int xsection_get_boot_mode_id_ext(u8 *buf);
+int xsection_set_boot_mode_id_ext(u8 *value);
+
 int xsection_get_devconf(u16 *buf);
 int xsection_set_devconf(u16 *value);
 
-- 
2.7.4

