From 38e56c22c5f7c52cddde044bfc45989864e4e77a Mon Sep 17 00:00:00 2001
From: "Karfich, Oleg" <Oleg.Karfich@wago.com>
Date: Tue, 11 Feb 2020 15:57:12 +0000
Subject: [PATCH] drivers: xsection: add blp_uii field as device-parameter

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 drivers/misc/xsection.c | 44 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 44 insertions(+)

diff --git a/drivers/misc/xsection.c b/drivers/misc/xsection.c
index cafda6f..894157b 100644
--- a/drivers/misc/xsection.c
+++ b/drivers/misc/xsection.c
@@ -37,6 +37,7 @@ struct xsection {
 	struct xsection_item ethaddr1;
 	struct xsection_item ethaddr2;
 	struct xsection_item blidx;
+	struct xsection_item blp_uii;
 };
 
 static struct xsection *xsection;
@@ -267,6 +268,41 @@ static int xsection_param_set_blidx(struct param_d *p, void *priv)
 	return ret < 0 ? ret : 0;
 }
 
+static int xsection_param_get_blp_uii(struct param_d *p, void *priv)
+{
+	struct xsection_item *xitem = priv;
+	int ret;
+	u8 buf[40];
+
+	ret = xsection_access_item(xitem, &buf, O_RDONLY);
+	if (ret < 0)
+		return ret;
+
+	/* replace last byte with terminating null byte to avoid garbage in the
+	 * string device parameter.
+	 */
+	buf[39] = '\0';
+
+	free(xitem->value_param);
+	xitem->value_param = basprintf("%s", buf);
+
+	return 0;
+}
+
+static int xsection_param_set_blp_uii(struct param_d *p, void *priv)
+{
+	struct xsection_item *xitem = priv;
+	int ret;
+
+	if (!strcmp(xitem->value_param, "") || strlen(xitem->value_param) != 39)
+		return -EINVAL;
+
+	/* we need to set a <EOT> sign at the end per definition */
+	xitem->value_param[39] = 0x04;
+	ret = xsection_access_item(xitem, xitem->value_param, O_WRONLY);
+
+	return ret < 0 ? ret : 0;
+}
 static int xsection_add_dev_param(struct device_node *np,
 				struct xsection *xsection,
 				struct xsection_item *xitem, char *param_name,
@@ -369,6 +405,14 @@ static int xsection_probe(struct device_d *dev)
 	if (ret)
 		goto out_unreg;
 
+	/* blp_uii is currently only used in pfc200 adv devices. So it's ok when
+	 * no "blp_uii" partition is found.
+	 */
+	xsection_add_dev_param(np, xsection, &xsection->blp_uii,
+					"blp_uii",
+					 xsection_param_set_blp_uii,
+					 xsection_param_get_blp_uii);
+
 	return 0;
 
 out_unreg:
-- 
2.7.4

