From 5094b3f988331a7820ccdb8f34374a0c4d20cde4 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Wed, 29 Apr 2020 12:12:03 +0200
Subject: [PATCH] eeprom: at24: consider EPROBE_DEFER return values

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 drivers/eeprom/at24.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/eeprom/at24.c b/drivers/eeprom/at24.c
index 4fae3a1..ffac69e 100644
--- a/drivers/eeprom/at24.c
+++ b/drivers/eeprom/at24.c
@@ -476,6 +476,12 @@ static int at24_probe(struct device_d *dev)
 		enum of_gpio_flags flags;
 		at24->wp_gpio = of_get_named_gpio_flags(dev->device_node,
 				"wp-gpios", 0, &flags);
+
+		if (at24->wp_gpio == -EPROBE_DEFER) {
+			err = at24->wp_gpio;
+			goto err_probe_defer;
+		}
+
 		if (gpio_is_valid(at24->wp_gpio)) {
 			at24->wp_active_low = flags & OF_GPIO_ACTIVE_LOW;
 			gpio_request(at24->wp_gpio, "eeprom-wp");
@@ -515,6 +521,7 @@ err_clients:
 
 	if (gpio_is_valid(at24->wp_gpio))
 		gpio_free(at24->wp_gpio);
+err_probe_defer:
 	kfree(at24->writebuf);
 err_device_name:
 	kfree(at24);
-- 
2.7.4

