From 63d7b7c0f02c1a2c1b34727246e2661b04cfe58f Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Thu, 12 Mar 2020 15:10:18 +0100
Subject: [PATCH] add new version w03.04.00_11

Signed-off-by: Christian Hohnstaedt <Christian.Hohnstaedt@wago.com>
---
 localversion-wago | 1 +
 1 file changed, 1 insertion(+)
 create mode 100644 localversion-wago

diff --git a/localversion-wago b/localversion-wago
new file mode 100644
index 00000000000..84a9c98c7d5
--- /dev/null
+++ b/localversion-wago
@@ -0,0 +1 @@
+w03.04.00_11
-- 
2.17.1

