--- syslog-ng-3.20.1/lib/host-resolve.c	2019-02-15 08:35:57.000000000 +0100
+++ syslog-ng-3.20.1/lib/host-resolve.c	2019-03-28 16:18:06.115548168 +0100
@@ -28,6 +28,7 @@
 #include "tls-support.h"
 #include "compat/socket.h"
 
+#include <netinet/in.h>
 #include <arpa/inet.h>
 #include <netdb.h>
 #include <string.h>
@@ -146,6 +147,12 @@
   hints.ai_protocol = 0;
   hints.ai_flags = AI_V4MAPPED | AI_ADDRCONFIG;

+  struct in_addr v4localhost;
+  if (    (g_ascii_strncasecmp(name, "localhost", 10) == 0)
+       || (g_ascii_strncasecmp(name, "::1", 4) == 0)
+       || ((inet_aton(name, &v4localhost) != 0) && (ntohl(v4localhost.s_addr) > 0x7f000000) && (ntohl(v4localhost.s_addr) < 0x7fffffff)))
+    hints.ai_flags = AI_V4MAPPED;
+
   if (getaddrinfo(name, NULL, &hints, &res) == 0)
     {
       /* we only use the first entry in the returned list */
