Index: ipwatchd-1.2.1/src/scripts/ipwatchd-script
===================================================================
--- ipwatchd-1.2.1.orig/src/scripts/ipwatchd-script	2011-07-24 14:39:03.000000000 +0000
+++ ipwatchd-1.2.1/src/scripts/ipwatchd-script	2014-06-27 14:15:19.795991817 +0000
@@ -20,5 +20,15 @@
 		--message "MAC address $MAC causes IP conflict with address $IP set on interface $DEVICE"
 fi
 
+# Create entry in /var/log/wago/wagolog.log.
+# Decode log with cat /var/log/wago/wagolog.log | decodeid.
+if [ -x /usr/bin/eventmsg ]; then
+	if [ $DEVICE = "br0" ]; then
+       		eventmsg 0x00000005
+	elif [ $DEVICE = "br1" ]; then
+       		eventmsg 0x00000006
+	fi
+fi
+
 exit 0
 
