Index: dropbear-2017.75/options.h
===================================================================
--- dropbear-2017.75.orig/options.h
+++ dropbear-2017.75/options.h
@@ -300,7 +300,7 @@ Homedir is prepended unless path begins
  * OpenSSH), set the path below. If the path isn't defined, sftp will not
  * be enabled */
 #ifndef SFTPSERVER_PATH
-#define SFTPSERVER_PATH "/usr/libexec/sftp-server"
+#define SFTPSERVER_PATH "/usr/sbin/sftp-server"
 #endif
 
 /* This is used by the scp binary when used as a client binary. If you're
