From: Marc Kleine-Budde <mkl@pengutronix.de>
Date: Fri, 8 Jan 2010 20:10:56 +0100
Subject: [PATCH] buildmacros: add --tag=CC to libtool invocation

Signed-off-by: Marc Kleine-Budde <mkl@pengutronix.de>
---
 include/buildmacros | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/buildmacros b/include/buildmacros
index 9e81bb83d609..805ef64d8d57 100644
--- a/include/buildmacros
+++ b/include/buildmacros
@@ -54,10 +54,10 @@ LIBNAME = $(basename $(LTLIBRARY))
 LTOBJECTS = $(OBJECTS:.o=.lo)
 LTVERSION = $(LT_CURRENT):$(LT_REVISION):$(LT_AGE)
 
-LTLINK = $(LIBTOOL) --mode=link $(CC)
+LTLINK = $(LIBTOOL) --tag=CC --mode=link $(CC)
 LTEXEC = $(LIBTOOL) --mode=execute
 LTINSTALL = $(LIBTOOL) --mode=install $(INSTALL)
-LTCOMPILE = $(LIBTOOL) --mode=compile $(CCF)
+LTCOMPILE = $(LIBTOOL) --tag=CC --mode=compile $(CCF)
 
 ifeq ($(ENABLE_SHARED),yes)
 LTLDFLAGS += -rpath $(PKG_LIB_DIR)
