Index: acl-2.2.52/include/install-sh
===================================================================
--- acl-2.2.52.orig/include/install-sh
+++ acl-2.2.52/include/install-sh
@@ -56,7 +56,7 @@ _chown ()
 {
     _st=255
     if [ $# -eq 3 ] ; then
-	chown $1:$2 $3
+	chown "$1":"$2" $3
 	_st=$?
 	if [ $_st -ne 0 ] ; then
 	    if [ $REAL_UID != '0' ] ; then
@@ -93,7 +93,7 @@ DIRMODE=755
 FILEMODE=644
 OWNER=`id -u`
 GROUP=`id -g`
-REAL_UID=$OWNER
+REAL_UID="$OWNER"
 
 # default is to install and don't append manifest
 INSTALL=true
@@ -115,7 +115,7 @@ fi
 
 [ -n "$DIST_ROOT" -a $REAL_UID -ne 0 ] && CHOWN=true
 
-while getopts "Dcm:d:S:o:g:T:" c $*
+while getopts "Dcm:d:S:o:g:T:" c "$@"
 do
    case $c in
    c)
@@ -168,10 +168,10 @@ then
     fi
     if [ $status -eq 0 ]
     then
-	$CHOWN $OWNER $GROUP $dir
+	$CHOWN "$OWNER" "$GROUP" $dir
 	status=$?
     fi
-    $MANIFEST d $DIRMODE $OWNER $GROUP ${dir#$DIST_ROOT}
+    $MANIFEST d $DIRMODE "$OWNER" "$GROUP" ${dir#$DIST_ROOT}
 elif $Sflag
 then
     #
@@ -217,7 +217,7 @@ then
 			install_name=$target/$solib
 			$CP $solib $install_name
 			status=$?
-			$MANIFEST f $FILEMODE $OWNER $GROUP $HERE/$solib ${install_name#$DIST_ROOT}
+			$MANIFEST f $FILEMODE "$OWNER" "$GROUP" $HERE/$solib ${install_name#$DIST_ROOT}
 			break
 		fi
 	done
@@ -268,7 +268,7 @@ then
 	install_name=$target/$old_library
 	$CP $old_library $install_name
 	status=$?
-	$MANIFEST f $FILEMODE $OWNER $GROUP $HERE/$old_library ${install_name#$DIST_ROOT}
+	$MANIFEST f $FILEMODE "$OWNER" "$GROUP" $HERE/$old_library ${install_name#$DIST_ROOT}
 	;;
     *)
 	echo "$prog: -T $lt_install invalid"
@@ -281,7 +281,7 @@ then
 	if [ $status -eq 0 ]
 	then
 		$CHMOD $FILEMODE $install_name
-		$CHOWN $OWNER $GROUP $install_name
+		$CHOWN "$OWNER" "$GROUP" $install_name
 	fi
 	;;
     esac
@@ -310,19 +310,19 @@ else
 		status=$?
 		if [ $status -eq 0 ]
 		then
-		    $CHOWN $OWNER $GROUP $dir/$f
+		    $CHOWN "$OWNER" "$GROUP" $dir/$f
 		    status=$?
 		fi
-		$MANIFEST f $FILEMODE $OWNER $GROUP $HERE/$f ${dir#$DIST_ROOT}/$f
+		$MANIFEST f $FILEMODE "$OWNER" "$GROUP" $HERE/$f ${dir#$DIST_ROOT}/$f
 	    else
 		$CHMOD $FILEMODE $dir
 		status=$?
 		if [ $status -eq 0 ]
 		then
-		    $CHOWN $OWNER $GROUP $dir
+		    $CHOWN "$OWNER" "$GROUP" $dir
 		    status=$?
 		fi
-		$MANIFEST f $FILEMODE $OWNER $GROUP $HERE/$dir ${dir#$DIST_ROOT}
+		$MANIFEST f $FILEMODE "$OWNER" "$GROUP" $HERE/$dir ${dir#$DIST_ROOT}
 	    fi
 	fi
     else
@@ -352,10 +352,10 @@ else
 		status=$?
 		if [ $status -eq 0 ]
 		then
-		    $CHOWN $OWNER $GROUP $dir/$f
+		    $CHOWN "$OWNER" "$GROUP" $dir/$f
 		    status=$?
 		fi
-		$MANIFEST f $FILEMODE $OWNER $GROUP $HERE/$f ${dir#$DIST_ROOT}/$f
+		$MANIFEST f $FILEMODE "$OWNER" "$GROUP" $HERE/$f ${dir#$DIST_ROOT}/$f
 	    fi
 	    [ $status -ne 0 ] && break
 	done
