From: Jobst Wellensiek <jobst.wellensiek@wago.com>
Date: Thu, 29 Apr 2021 10:12:06 +0200
Subject: [PATCH] fix memory leak Unified Automation patch

---
 src/uaserver/uaserverc/uaserver_monitoreditemmanager.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/uaserver/uaserverc/uaserver_monitoreditemmanager.c b/src/uaserver/uaserverc/uaserver_monitoreditemmanager.c
index bc398084c62a..811570ef2980 100644
--- a/src/uaserver/uaserverc/uaserver_monitoreditemmanager.c
+++ b/src/uaserver/uaserverc/uaserver_monitoreditemmanager.c
@@ -288,7 +288,7 @@ OpcUa_InitializeStatus(OpcUa_Module_Server, "UaServer_NewItemValue");
             }
             /* copy current value for next interval compare */
             OpcUa_DataValue_Clear(&a_pMonitoredItemData->LastDataValue);
-            OpcUa_DataValue_CopyTo(&currentValue, &a_pMonitoredItemData->LastDataValue);
+            a_pMonitoredItemData->LastDataValue = currentValue;
         }
         else /* do not discard oldest */
         {
@@ -317,7 +317,7 @@ OpcUa_InitializeStatus(OpcUa_Module_Server, "UaServer_NewItemValue");
 
             /* copy current value for next interval compare */
             OpcUa_DataValue_Clear(&a_pMonitoredItemData->LastDataValue);
-            OpcUa_DataValue_CopyTo(&currentValue, &a_pMonitoredItemData->LastDataValue);
+            a_pMonitoredItemData->LastDataValue = currentValue;
         }
 
         uStatus = UaServer_ItemChanged((UaServer_MonitoredItem*)a_pMonitoredItemData);
