Index: protobuf-3.12.3/cmake/protobuf-config.cmake.in
===================================================================
--- protobuf-3.12.3.orig/cmake/protobuf-config.cmake.in
+++ protobuf-3.12.3/cmake/protobuf-config.cmake.in
@@ -99,7 +99,7 @@ function(protobuf_generate)
   foreach(_proto ${protobuf_generate_PROTOS})
     get_filename_component(_abs_file ${_proto} ABSOLUTE)
     get_filename_component(_abs_dir ${_abs_file} DIRECTORY)
-    get_filename_component(_basename ${_proto} NAME_WLE)
+    get_filename_component(_basename ${_proto} NAME_WE)
 
     set(_suitable_include_found FALSE)
     foreach(DIR ${_protobuf_include_path})
