From a43db0e9f8dec9cd1d9e4a075cba197741f6a285 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Tue, 7 Jul 2015 13:27:59 +0200
Subject: [PATCH] can: ti: disable priority table lookup

---
 drivers/net/can/ti_hecc.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/net/can/ti_hecc.c b/drivers/net/can/ti_hecc.c
index 70c577a..750a0411c 100644
--- a/drivers/net/can/ti_hecc.c
+++ b/drivers/net/can/ti_hecc.c
@@ -867,7 +867,12 @@ static int ti_hecc_open(struct net_device *ndev)
 	struct ti_hecc_priv *priv = netdev_priv(ndev);
 	int err;
 
-	err = request_irq(ndev->irq, ti_hecc_interrupt, IRQF_SHARED | IRQF_THREAD_TBL_LOOKUP,
+	/*
+		temporaly fix. can stack generates timeout if we are using
+		the look-up table to set priority
+	*/
+	//err = request_irq(ndev->irq, ti_hecc_interrupt, IRQF_SHARED | IRQF_THREAD_TBL_LOOKUP,
+	err = request_irq(ndev->irq, ti_hecc_interrupt, IRQF_SHARED,
 			ndev->name, ndev);
 	if (err) {
 		netdev_err(ndev, "error requesting interrupt\n");
-- 
2.7.4

