From 76b453685089299cff98473732b333d499292186 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Mon, 3 Jul 2017 16:20:20 +0200
Subject: [PATCH 078/261] net: dsa: ksz8863: fix compiler warning

---
 drivers/net/dsa/ksz8863.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/net/dsa/ksz8863.c b/drivers/net/dsa/ksz8863.c
index 2234d09..1e90f34 100644
--- a/drivers/net/dsa/ksz8863.c
+++ b/drivers/net/dsa/ksz8863.c
@@ -276,7 +276,6 @@ struct ksz8863_info *of_get_ksz8863_pdata(struct device *dev)
 {
 	struct device_node *np = dev->of_node;
 	struct pinctrl *pinctrl;
-	int ret;
 
 	ksz_info = devm_kzalloc(dev, sizeof(*ksz_info), GFP_KERNEL);
 	if (!ksz_info)
-- 
2.7.4

