From 1180b3a01e17f24fcb5c250347ec5eeba2c726d0 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Fri, 21 Jul 2017 00:44:47 +0000
Subject: [PATCH 088/261] spi: kbus: use gpio cansleep for reset to prevent
 dumps

---
 drivers/spi/spi-kbus.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/spi/spi-kbus.c b/drivers/spi/spi-kbus.c
index df83390..0991fbe 100644
--- a/drivers/spi/spi-kbus.c
+++ b/drivers/spi/spi-kbus.c
@@ -868,9 +868,9 @@ static int kbus_trig_reset(struct kbus_drv_data *kdrvdata)
 		return -EINVAL;
 
 	/* reset kbus slave cpu (Infineon XE164) */
-	gpio_set_value(kdrvdata->gpio_nrst, kdrvdata->nrst_alow);
+	gpio_set_value_cansleep(kdrvdata->gpio_nrst, kdrvdata->nrst_alow);
 	udelay(100);
-	gpio_set_value(kdrvdata->gpio_nrst, !kdrvdata->nrst_alow);
+	gpio_set_value_cansleep(kdrvdata->gpio_nrst, !kdrvdata->nrst_alow);
 
 	pr_info("PFCxxx-KBUS: Kbus Slave CPU Reset (gpio %d, active %s) ... Done.\n",
 		kdrvdata->gpio_nrst, kdrvdata->nrst_alow ? "low" : "high");
-- 
2.7.4

