From 736307dd2fa25c303f381f3ae596f96c3fabd021 Mon Sep 17 00:00:00 2001
From: Oliver Schildan <oliver.schildan@wago.com>
Date: Thu, 16 Nov 2017 08:54:45 +0000
Subject: [PATCH] 3G-modem: fix 3G-modem reset copy and past error

---
 arch/arm/mach-omap2/wsysinit.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/arch/arm/mach-omap2/wsysinit.c b/arch/arm/mach-omap2/wsysinit.c
index a8dedd6..f111a5e 100644
--- a/arch/arm/mach-omap2/wsysinit.c
+++ b/arch/arm/mach-omap2/wsysinit.c
@@ -191,8 +191,7 @@ static int wsysinit_init_probe_dt(struct platform_device *pdev)
 		if (wsysinitset.modem_reset)
 			modem_reset(&wsysinitset);
 
-		/* devm_gpio_free(&pdev->dev, wsysinitset.dp_gpio_rst); */
-	} else if (wsysinitset.dp_reset) {
+	} else if (wsysinitset.modem_reset) {
 		dev_err(&pdev->dev, "failed to find modem reset signal!\n");
 		return -1;
 	}
-- 
2.7.4

