From 4cdc938bf491ac434975a9a3750fc1193848f3dc Mon Sep 17 00:00:00 2001
From: Peter Galka <peter.galka@wago.com>
Date: Thu, 1 Feb 2018 10:57:50 +0100
Subject: [PATCH 151/261] drivers: cpsw: fix host port initial ale entry

---
 drivers/net/ethernet/ti/cpsw.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/ti/cpsw.c b/drivers/net/ethernet/ti/cpsw.c
index b91a583..e8c40a4 100644
--- a/drivers/net/ethernet/ti/cpsw.c
+++ b/drivers/net/ethernet/ti/cpsw.c
@@ -1293,7 +1293,7 @@ static void cpsw_init_host_port(struct cpsw_priv *priv)
 
 	if (!cpsw->data.dual_emac) {
 		cpsw_ale_add_ucast(cpsw->ale, priv->mac_addr, HOST_PORT_NUM,
-				   0, 0);
+				   ALE_SECURE, 0);
 		cpsw_ale_add_mcast(cpsw->ale, priv->ndev->broadcast,
 				   ALE_PORT_HOST, 0, 0, ALE_MCAST_FWD_2);
 	}
-- 
2.7.4

