From 4058485615c8f5462c354c94e3ef9fae5dd46520 Mon Sep 17 00:00:00 2001
From: Boerje Sewing <Boerje.Sewing@who-ing.de>
Date: Wed, 21 Feb 2018 10:21:32 +0100
Subject: [PATCH] ksz8863: add error message when held in reset.

Signed-off-by: Boerje Sewing <Boerje.Sewing@who-ing.de>
---
 drivers/net/dsa/ksz8863.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/net/dsa/ksz8863.c b/drivers/net/dsa/ksz8863.c
index 45d7848..878ca5b 100644
--- a/drivers/net/dsa/ksz8863.c
+++ b/drivers/net/dsa/ksz8863.c
@@ -211,9 +211,13 @@ static const char *ksz8863_probe(struct device *dsa_dev,
 	if ((ret = ksz8863_smi_nread(ChipID0, chipid, 2)))
 		return NULL;
 
-	if (chipid[0] == KSZ8863_FAMILY_ID &&
-	    (chipid[1] >> 4) == KSZ8863_CHIP_ID)
-		dev_info(ksz_info->dev, "Detected Family ID 0x%x, Chip ID 0x%x\n", chipid[0], chipid[1] >> 4);
+	if (chipid[0] != KSZ8863_FAMILY_ID ||
+	    (chipid[1] >> 4) != KSZ8863_CHIP_ID) {
+		dev_err(ksz_info->dev, "could not detect ksz8863 switch.\n");
+		return NULL;
+	}
+
+	dev_info(ksz_info->dev, "Detected Family ID 0x%x, Chip ID 0x%x\n", chipid[0], chipid[1] >> 4);
 
 	if (ksz_info->disable_internal_ldo) {
                 if ((ret = ksz8863_smi_nread(Internal1v8ldoControl, &v, 1)))
-- 
2.7.4

