From 0e5aa968d824c2ee16294561e896d438e28ad48f Mon Sep 17 00:00:00 2001
From: Andreas Schmidt <andreas.schmidt@wago.com>
Date: Tue, 24 Jul 2018 15:04:50 +0200
Subject: [PATCH 221/261] pnsk: add if_pn.h to exported kernel-headers

Signed-off-by: Andreas Schmidt <andreas.schmidt@wago.com>
---
 include/uapi/linux/Kbuild | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/uapi/linux/Kbuild b/include/uapi/linux/Kbuild
index ccdd8d5..bd9bb25 100644
--- a/include/uapi/linux/Kbuild
+++ b/include/uapi/linux/Kbuild
@@ -182,6 +182,7 @@ header-y += if_macsec.h
 header-y += if_packet.h
 header-y += if_phonet.h
 header-y += if_plip.h
+header-y += if_pn.h
 header-y += if_ppp.h
 header-y += if_pppol2tp.h
 header-y += if_pppox.h
-- 
2.7.4

