From b789c94404f5e05daf75d3493a445b36faa46472 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Tue, 8 Jan 2019 14:41:33 +0100
Subject: [PATCH 291/291] add new version w02.03.00_02

Signed-off-by: Heinrich Toews <heinrich.toews@wago.com>
---
 localversion-wago | 1 +
 1 file changed, 1 insertion(+)
 create mode 100644 localversion-wago

diff --git a/localversion-wago b/localversion-wago
new file mode 100644
index 0000000..ae30e76
--- /dev/null
+++ b/localversion-wago
@@ -0,0 +1 @@
+-w02.03.00_02
-- 
2.7.4

