From 64f9e26a9db2a0c68a35c094df433fc47ae99561 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Wed, 9 Jan 2019 15:18:44 +0100
Subject: [PATCH 294/294] add new version w02.03.00_02+1

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 localversion-wago | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-wago b/localversion-wago
index ae30e76..cf9f66c 100644
--- a/localversion-wago
+++ b/localversion-wago
@@ -1 +1 @@
--w02.03.00_02
+-w02.03.00_02+1
-- 
2.7.4

