Index: openvpn-2.3.13/src/openvpn/options.c
===================================================================
--- openvpn-2.3.13.orig/src/openvpn/options.c
+++ openvpn-2.3.13/src/openvpn/options.c
@@ -5698,7 +5698,15 @@ add_option (struct options *options,
     {
         VERIFY_PERMISSION(OPT_P_GENERAL);
         options->log = true;
-        redirect_stdout_stderr(p[1], false);
+
+        // To make openvpn compatible with
+        // logrotate's copytruncate option,
+        // the log file has to be always
+        // appended to rather than truncated.
+        // Effectively, "log" configuration
+        // option is mapped to "log-append"
+        // here.
+        redirect_stdout_stderr(p[1], true);
     }
     else if (streq(p[0], "suppress-timestamps") && !p[1])
     {
