From: Michael Olbrich <m.olbrich@pengutronix.de>
Date: Mon, 11 Aug 2014 12:28:49 +0200
Subject: [PATCH] Configure: don't ask dpkg-buildflags for more flags

Signed-off-by: Michael Olbrich <m.olbrich@pengutronix.de>
---
 Configure | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Configure b/Configure
index 9f58145ef000..4b6f13ee238c 100755
--- a/Configure
+++ b/Configure
@@ -134,7 +134,7 @@ my $clang_devteam_warn = "-Wno-unused-parameter -Wno-missing-field-initializers
 my $warn_make_depend = 0;
 
 # There are no separate CFLAGS/CPPFLAGS/LDFLAGS, set everything in CFLAGS
-my $debian_cflags = `dpkg-buildflags --get CFLAGS` . `dpkg-buildflags --get CPPFLAGS` . `dpkg-buildflags --get LDFLAGS` . "-Wa,--noexecstack -Wall";
+my $debian_cflags = "-g -O2 -Wformat -Werror=format-security " . "-Wa,--noexecstack -Wall";
 $debian_cflags =~ s/\n/ /g;
 
 my $strict_warnings = 0;
