From: Markus Niebel <Markus.Niebel@tq-group.com>
Date: Fri, 31 Mar 2017 11:46:14 +0200
Subject: [PATCH] mtd-probe: fix compilation

Signed-off-by: Markus Niebel <Markus.Niebel@tq-group.com>
---
 src/mtd_probe/mtd_probe.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/mtd_probe/mtd_probe.h b/src/mtd_probe/mtd_probe.h
index 2a37ede578b3..961176a475a2 100644
--- a/src/mtd_probe/mtd_probe.h
+++ b/src/mtd_probe/mtd_probe.h
@@ -17,6 +17,7 @@
  * Boston, MA  02110-1301  USA
  */
 
+#include <stdint.h>
 #include <mtd/mtd-user.h>
 
 /* Full oob structure as written on the flash */
