Index: azure-iot-sdk-c-1.2.9/umqtt/src/mqtt_client.c
===================================================================
--- azure-iot-sdk-c-1.2.9.orig/umqtt/src/mqtt_client.c
+++ azure-iot-sdk-c-1.2.9/umqtt/src/mqtt_client.c
@@ -1230,6 +1230,7 @@ int mqtt_client_disconnect(MQTT_CLIENT_H
 
             close_connection(mqtt_client);
             clear_mqtt_options(mqtt_client);
+            mqtt_client->xioHandle = NULL;
             result = 0;
         }
     }
