From 42f36c2a68f7b7f2bb593135334d3515d7dd0849 Mon Sep 17 00:00:00 2001
From: User <user@wago-vm.wago.local>
Date: Wed, 23 Oct 2019 15:45:52 +0200
Subject: [PATCH] Add error value for fatal error in loop forever

Signed-off-by: Benjamin Czytrich <benjamin.czytrich@wago.com>
---
 lib/mosquitto.c | 2 ++
 lib/mosquitto.h | 1 +
 2 files changed, 3 insertions(+)

diff --git a/lib/mosquitto.c b/lib/mosquitto.c
index d1510bc..25574b5 100644
--- a/lib/mosquitto.c
+++ b/lib/mosquitto.c
@@ -374,6 +374,8 @@ const char *mosquitto_strerror(int mosq_errno)
 			return "Keepalive timeout.";
 		case MOSQ_ERR_MALFORMED_UTF8:
 			return "Malformed UTF-8";
+		case MOSQ_ERR_LOOP_FOREVER_EXITED:
+			return "Loop forever exited.";
 		default:
 			return "Unknown error.";
 	}
diff --git a/lib/mosquitto.h b/lib/mosquitto.h
index 8f23948..b22d289 100644
--- a/lib/mosquitto.h
+++ b/lib/mosquitto.h
@@ -87,6 +87,7 @@ enum mosq_err_t {
 	MOSQ_ERR_MALFORMED_UTF8 = 18,
 	MOSQ_ERR_KEEPALIVE = 19,
 	MOSQ_ERR_LOOKUP = 20,
+	MOSQ_ERR_LOOP_FOREVER_EXITED = 21,
 };
 
 /* Error values */
-- 
2.7.4

