--- a/modules/pam_pwhistory/opasswd.c	2020-05-28 13:13:20.042526458 +0200
+++ b/modules/pam_pwhistory/opasswd.c	2020-05-28 13:13:06.154630816 +0200
@@ -55,6 +55,7 @@
 #include <crypt.h>
 #endif
 
+#include <security/_pam_macros.h> // x_strdup
 #include <security/pam_ext.h>
 #include <security/pam_modules.h>
 
@@ -102,18 +103,28 @@
 static int
 compare_password(const char *newpass, const char *oldpass)
 {
-  char *outval;
+  int retval;
+  char *outval = NULL;
 #ifdef HAVE_CRYPT_R
-  struct crypt_data output;
+  struct crypt_data *output;
 
-  output.initialized = 0;
-
-  outval = crypt_r (newpass, oldpass, &output);
+  output = calloc(1, sizeof(*output));
+  if(output != NULL)
+  {
+    outval = x_strdup(crypt_r (newpass, oldpass, output));
+    memset(output, '\0', sizeof(*output));
+    free(output);
+  }
 #else
-  outval = crypt (newpass, oldpass);
+  outval = x_strdup(crypt (newpass, oldpass));
 #endif
 
-  return outval != NULL && strcmp(outval, oldpass) == 0;
+  retval = outval != NULL && strcmp(outval, oldpass) == 0;
+
+  if(outval)
+    free(outval);
+
+  return retval;
 }
 
 /* Check, if the new password is already in the opasswd file.  */
